pallet/dex/src/lib.rs

```diff
@@ -0,0 +1,568 @@
+			block_number: u64,
+		},
+		OrderCreated {
+			order_index: u64,
+			order: OrderOf<T>,
+		},
+		OrderTaken {
+			account: T::AccountId,
+			order_index: u64,
+			order: OrderOf<T>,
+		},
+		OrderCanceled {
+			order_index: u64,
+		},
+		Deposited {
+			asset_id: u32,
+			amount: u128,
+		},
+
+		Withdrawed {
+			asset_id: u32,
+			amount: u128,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		OrderIndexOverflow,
+		InvalidOrderIndex,
+		InsufficientBalance,
+		NotOwner,
+		AssetIdNotInTokenIndex,
+		AssetIdNotInTokenInfoes,
+		TokenBalanceOverflow,
+		WithdrawBalanceMustKeepOrderSellAmount,
+		UserAssetNotExist,
+		PairOrderNotFound,
+		PairAssetIdMustNotEqual,
+		NotEnoughBalance,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight({0})]
+		#[pallet::call_index(0)]
+		pub fn deposit(
+			origin: OriginFor<T>,
+			asset_id: u32,
+			amount: u128,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				TokenIndex::<T>::contains_key(asset_id),
+				Error::<T>::AssetIdNotInTokenIndex
+			);
+
+			<T::Fungibles as Mutate<T::AccountId>>::transfer(
+				asset_id,
+				&who,
+				&Self::account_id(),
+				amount,
+				Preservation::Expendable,
+			)?;
+
+			let mut info = TokenInfo::default();
+			if UserTokenInfoes::<T>::contains_key(who.clone(), asset_id) {
+				info = UserTokenInfoes::<T>::get(who.clone(), asset_id);
+				info.amount = info
+					.amount
+					.checked_add(amount)
+					.ok_or(Error::<T>::TokenBalanceOverflow)?;
+			} else {
+				info.amount = amount;
+			}
+
+			UserTokenInfoes::<T>::insert(who, asset_id, info);
+
+			Self::deposit_event(Event::Deposited { asset_id, amount });
+
+			Ok(().into())
+		}
+
+		#[pallet::weight({1})]
+		#[pallet::call_index(1)]
+		pub fn withdraw(
+			origin: OriginFor<T>,
+			asset_id: u32,
+			amount: u128,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				TokenIndex::<T>::contains_key(asset_id),
+				Error::<T>::AssetIdNotInTokenIndex
+			);
+
+			ensure!(
+				UserTokenInfoes::<T>::contains_key(who.clone(), asset_id),
+				Error::<T>::AssetIdNotInTokenInfoes
+			);
+
+			let mut info = UserTokenInfoes::<T>::get(who.clone(), asset_id);
+			info.amount = info
+				.amount
+				.checked_sub(amount)
+				.ok_or(Error::<T>::NotEnoughBalance)?;
+
+			<T::Fungibles as Mutate<T::AccountId>>::transfer(
+				asset_id,
+				&Self::account_id(),
+				&who,
+				amount,
+				Preservation::Expendable,
+			)?;
+
+			UserTokenInfoes::<T>::insert(who, asset_id, info);
+
+			Self::deposit_event(Event::Withdrawed { asset_id, amount });
+			Ok(().into())
+		}
+
+		#[pallet::weight({2})]
+		#[pallet::call_index(2)]
+		pub fn make_order(
+			origin: OriginFor<T>,
+			asset_id_1: u32,
+			asset_id_2: u32,
+			offered_amount: u128,
+			requested_amount: u128,
+			order_type: OrderType,
+		) -> DispatchResultWithPostInfo {
```

Comment:
```suggestion
		) -> DispatchResultWithPostInfo {
		let (asset_id_1, asset_id_2) = if asset_id_1 > asset_id_2 {
		   (asset_id_2, asset_id_1)
		} else { (asset_id_1, asset_id_2) };
```

We should maintain this invariant elsewhere, otherwise it's harder to match `(1u32, 2u32)` since there can be orders on the book for `(2u32, 1u32)`.

Reply:
ok